// pathr/lib/routes.js
'use strict';

var utils = require('./utils.js');

/**
 *	Handle a request
 *	Match the path against the regexps we have
 *	
 *	@param {String} method
 *	@param {String} path
 *	@param {Function} callback
 *	@param {Object} options
 */
var Route = function(method, path, callback, options) {
	options = this.options = options || {};
	this.path = path;
	this.method = method;
	this.callback = callback;
	this.keys = [];
	this.regexp = utils.pathtoRegexp(path, this.keys);
	if (!this.regexp) return false;
};


/**
 *	Match a path agains the registered paths
 *	Make an object like array on this.params
 *	@param {String} path
 */
Route.prototype.match = function(path) {
	var keys = this.keys,
		params = this.params = [],
		m = this.regexp.exec(path);
	
	if (!m) return false;

	for (var i = 1, len = m.length; i<len; i++) {
		var key = keys[i-1],
			val = 'string' == typeof m[i] ? decodeURIComponent(m[i]) : m[i];

		if (key) {
			params[key.name] = val;
		} else {
			params.push(val);
		};
	};
	
	return params;
};

module.exports = Route;
